### ISSUE TITLE
CI plan for testing kyma upgrades

### ISSUE THREAD

**Issue by pbochynski** [kyma-project/kyma#2259]
**Description**

Prepare a CI plan that tests kyma upgrades from the latest release to the version created from pull request.

**Reasons**

The master branch should contain code that can be released and the version produced from this released can update previously installed one without issues.
We should know early if the upgrade process will not work.

**Acceptance criteria**
- before current version is installed (from master or pull request) previues version is installed (0.6)
- new check is created in github for kyma pull requests (optional - can be enabled if all components are upgradeable)
- kyma docs updated with documentation about Kyma upgrade

**Comment by PK85** [comment]
Probably three jobs need to be crated:
- presubmit job
- postsubmit job
- release job

**Comment by mszostok** [comment]
### TL;DR;

The upgradeability plan is executed on the **GKE** cluster. For now this plan **is failing on upgrade action.**

There are defined three jobs:
- presubmit for PR (runs automatically)
- postsubmit for master (runs automatically)
- and presubmit for releases (needs to be triggered manually)

_**NOTE:** Those jobs are not blocking the PR merges._

Follow-up task to fix the upgradeability plan:
- Service Catalog - TBD

**NOTE:** Before executing the upgradeability job for release you need to take some manual action. More information can be found in the details section.

---
### Details

<details><summary>Q: Why only GKE cluster?
<br> Click for answer</summary>
<p>
<br> 
The is the ability to run the upgradeability plan both on minikube and cluster. I choose to run upgradeability plan only on GKE. The reason is that this plan is mostly for ensuring that client will be able to upgrade his cluster and not the local installation. Additionally, on GKE cluster we are installing all components and in the local case, we skipping some of them e.g. monitoring. With the local approach, we will not test all Kyma functionality. 
</p>
</details>

<br>
<details><summary>Q: How the latest release is chosen?<br> Click for answer</summary>
<p>
<br> 
We are always getting the latest release of Kyma based on the SemVer and not the order on the Github. Draft and Prereleases are omitted.
</p>
</details>


<br>
<details><summary>Q: How the flow for PR/master looks like?<br> Click for answer</summary>
<p>
<br> 

The job is triggered **automatically** by Prow.

1. The GKE cluster is provisioned.
2. The lastest Kyma release is installed. 
3. Kyma is upgraded.
	3.1 The kyma installer from actual PR/master builds and pushed to gcr registry.
	3.2 The [**installer.yaml**](https://github.com/kyma-project/kyma/blob/master/installation/resources/installer.yaml) and [**installer-cr-cluster.yaml**](https://github.com/kyma-project/kyma/blob/master/installation/resources/installer-cr-cluster.yaml.tpl) YAMLs from actual PR/master are applied. YAML with overrides is omitted.
4. Kyma is tested. The [**testing.sh**](https://github.com/kyma-project/kyma/blob/master/installation/scripts/testing.sh) script from actual PR/master is executed.

</p>
</details>

<br>
<details><summary>Q: How the flow for release looks like?<br> Click for answer</summary>
<p>
<br> 

The job needs to be triggered **manually** by adding such comment: 
> /test pre-master-kyma-gke-upgradeability

The same flow we already have for the current release process, see: https://github.com/kyma-project/kyma/pull/2295#issuecomment-454361245


Flow for release:
1. The GKE cluster is provisioned.
2. The lastest Kyma release is installed. 
3. Kyma is upgraded.
	3.1 The **kyma-config-upgrade-cluster.yaml** artifact from current release is downloaded and applied.
4. Kyma is tested. The [**testing.sh**](https://github.com/kyma-project/kyma/blob/master/installation/scripts/testing.sh) script from actual release branch is executed.

The **kyma-config-upgrade-cluster.yaml** needs to be created and uploaded manually. 
This file needs to contain the:
- Deployment for Kyma Installer (with all required ClusterRoles etc.)
- Installation CR with the list of components to install
- All new overrides which are required since the previous release

I can see the possibility to automate this process.
First, we can generate file which will contain the Installer Deployment and the Installation CR
```
kyma/installation/scripts/concat-yamls.sh installer.yaml installer-cr-cluster.yaml.tpl \
| sed -e 's;image: eu.gcr.io/kyma-project/.*/installer:.*$;'"image: ${KYMA_INSTALLER_RELEASE_IMAGE_VERSION};" \
| sed -e "s/__VERSION__/0.0.1/g" \
| sed -e "s/__.*__//g" \
> kyma-config-upgrade-cluster.yaml
```

Then we need to create overrides diff based on files from the previous and current release. All new things should be added to the final **kyma-config-upgrade-cluster.yaml** file. 
This functionality can be added [here](https://github.com/kyma-project/kyma/blob/master/installation/scripts/release-generate-kyma-installer-artifacts.sh), where we already generating the **kyma-config-local.yaml** and **kyma-config-cluster.yaml** files.

**But IMO still this file needs to be validated by human.**
</p>
</details>

<br>
<details><summary>Q: How script can be improved in future?<br> Click for answer</summary>
<p>
<br> 

We can 
</p>
</details>
